### PR TITLE
swarm: add swarm integration

### LINKED ISSUE #87: Run Swarm Daemon
Status is going to require Swarm to operate higher functionalities, we'll need to have the ability to access Swarm as well

### PR DESCRIPTION
**Pull request by mchusovlianov**
- closes #87 

- it requires changes for safety multiple stops in hive 

- it runs with random key, it re-creates swarm with user selected key after Select and ReSelect calls

### CODE REVIEW

File: cmd/status/utils.go [hunk 1]
```diff
@@ -87,6 +87,10 @@ func testExportedAPI(t *testing.T, done chan struct{}) {
 			"pause node",
 			testStopResumeNode,
 		},
+		{
+			"swarm upload/download file",
+			testSwarmUploadFile,
+		},
 	}
```
Comment (farazdagi) [inline cmd/status/utils.go]: - Unnecessary, duplicate test
- What is being tested here that hasn't (or couldn't have) been tested in `swarm_test.go`?
- The `cmd/status/utils.go` is for tests in **exposed** methods i.e. methods that we expose in `library.go` - you haven't exposed any methods, no need to update `cmd/status/*` tests

File: geth/accounts.go [hunk 2]
```diff
@@ -153,6 +154,16 @@ func SelectAccount(address, password string) error {
 		return ErrWhisperIdentityInjectionFailure
 	}
 
+
+	swarmService, err := nodeManager.SwarmService()
+	if err != nil {
+		return err
+	}
+
+	if err := swarmService.InjectIdentity(accountKey.PrivateKey); err != nil {
```
Comment (farazdagi) [inline geth/accounts.go]: - it would have been better if we had sth like `swarmService.SwitchIdentity(pkey)` (or `SelectIdentity()`, or `SetIdentity`, or `SetServiceKey()`) method here, i.e. something that will not blindly copy what we have in Whisper (where identity injection makes sense). In swarm, the method forces swarm to do some changes - restart in our case - with base key/identity switched.
- again, all those if you think that proxy should exist, and you don't want to call `StartSwarm(pkey)` directly (I honestly don't see why)
- see my comments for `swarm.go`

File: geth/accounts.go [hunk 3]
```diff
@@ -201,6 +221,16 @@ func Logout() error {
 		return fmt.Errorf("%s: %v", ErrWhisperClearIdentitiesFailure, err)
 	}
 
+	swarmService, err := nodeManager.SwarmService()
+	if err != nil {
+		return err
+	}
+
+	err = swarmService.ClearIdentities()
```
Comment (farazdagi) [inline geth/accounts.go]: - again, why can't we stop swarm service directly with `swarmService.Stop()` w/o all those identity-related methods

File: geth/accounts_test.go [hunk 4]
```diff
@@ -300,6 +311,9 @@ func TestRecoverAccount(t *testing.T) {
 	if !whisperService.HasIdentity(crypto.ToECDSAPub(common.FromHex(pubKeyCheck))) {
 		t.Errorf("identity not injected into whisper: %v", err)
 	}
+	if !swarmService.HasIdentity(crypto.ToECDSAPub(common.FromHex(pubKeyCheck))) {
```
Comment (farazdagi) [inline geth/accounts_test.go]: - no need to check if identity exist in Swarm
- all those identity management were just copied from Whisper (where they make sense, as you can have multiple identities decrypting incoming traffic, in swarm single protocol identity is enough, isn't it?)
- checking that swarm service runs ok might be beneficial though

File: geth/accounts_test.go [hunk 5]
```diff
@@ -335,6 +355,9 @@ func TestAccountSelect(t *testing.T) {
 	if whisperService.HasIdentity(crypto.ToECDSAPub(common.FromHex(pubKey1))) {
 		t.Error("identity already present in whisper")
 	}
+	if swarmService.HasIdentity(crypto.ToECDSAPub(common.FromHex(pubKey1))) {
```
Comment (farazdagi) [inline geth/accounts_test.go]: - ditto (for this and all the `HasIdentity()` calls below

File: geth/node.go [hunk 6]
```diff
@@ -215,6 +221,18 @@ func activateShhService(stack *node.Node) error {
 	return nil
 }
 
+// activateShhService configures Whisper and adds it to the given node.
```
Comment (farazdagi) [inline geth/node.go]: - copy/paste survivor

File: geth/swarm.go [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+
+func NewSwarmService() *SwarmService {
+	service := &SwarmService{
+		keys: make(map[string]*ecdsa.PrivateKey),
+	}
+
+	return service
+}
+
+// activateEthService configures and registers the eth.Ethereum service with a given node.
```
Comment (farazdagi) [inline geth/swarm.go]: - fix comment
- consider renaming method, for ex `RunSwarmNode(pkey)` or sth, otherwise `swarm.Start()` and `swarm.StartSwarm()` are a bit confusing

File: geth/swarm.go [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+			return err
+		}
+
+		self.swarm = nil
+	}
+
+	return nil
+}
+
+func (self *SwarmService) Protocols() []p2p.Protocol {
```
Comment (farazdagi) [inline geth/swarm.go]: - purpose of introducing `Protocols()` and `APIs()` (below) methods?
- your custom `SwarmService`, that you introduced is supposed to be just a wrapper containing some config + `*swarm.Swarm`
- now, on why you've introduced these methods: to satisfy `node.Service` interface in https://github.com/status-im/status-go/pull/106/files#diff-dacdf432329ae450c554894d234b257dR225
- but then in addition to that you also needed to implement the whole `swarm_api_proxy.go`, again why?
- instead of producing all that code you could have rewrite swarm registration methods as:

```go
func activateSwarmService(stack *node.Node) error {
	serviceConstructor := func(ctx *node.ServiceContext) (node.Service, error) {
		swarmService := NewSwarmService()
		return swarmService.swarm, nil
	}
	if err := stack.Register(serviceConstructor); err != nil {
		return err
	}

	return nil
}
```

That's 2 lines of change! Note: you will need to init `*swarm.Swarm` in `NewSwarmService()` - even with randomly generated key, if necessary (basically, you need to call `StartSwarm(key)` in that constructor. 



To sum up: No need to introduce some crazy proxies and interfaces (which you don't need to satisfy semantically). Rely on `SwarmService.swarm` if you need to satisfy `node.Service` interface, no need to overextend interface to where it shouldn't belong.

File: geth/swarm_api_proxy.go [hunk 9]
```diff
@@ -0,0 +1,241 @@
+package geth
```
Comment (farazdagi) [inline geth/swarm_api_proxy.go]: - please remove this file
- in future, consider asking for advise before introducing something as dubious as this

File: geth/swarm.go (hunk 8 above)
Comment (mchusovlianov) [inline geth/swarm.go]: There is only one problem. We have to re-create swarm.Swarm after select account. I cannot find a way how to replace service in running node.Node. It needs for correct work functions from jailed context (web3.bzz.put).

Comment (farazdagi) [inline geth/swarm.go]: - well, if you need to manage `swarm.Swarm` object yourself (which you do in `StartSwarm()` via calling `NewSwarm()` and producing new instance of the service everytime), then probably you do not need to `activateSwarmService()` at all?
- At least there's little to no point passing `stack.Register()` anything that you need to replace on every account select (be it in form of proxy or direct `swarm.Swarm` instance)
- so, since you are already managing `swarm.Swarm` creation and swapping, then probably just not add it to stack at all?

Comment (farazdagi) [inline geth/swarm.go]: **Important:**
- The problem with "let's wrap everything into a proxy" is that we will need to keep proxy in sync with underlying implementation.
- Ideally, in future swarm will allow changing running private key on fly, which will simplify our code even further. And we even can make this happen, as nothing prevents us from modifying our fork of `go-ethereum`
- Until that, let's try to keep our swarm management code to the minimally possible, w/o reindexing public/private protocol swarm exposes via our proxies.

Basically, all we need is to start/stop service by feeding into it some parameters (like private key).

**Comment by farazdagi** [comment]
You've done marvelous job, thank you. We are almost there!

Can you please:
- rebase against the current `develop`. I've updated our fork of `go-ethereum` with your swarm fixes (both from upstream and unmerged), and pushed them to our `vendor` folder.
- squash every delta into a single commit
- force push your updated branch to this PR (so that I work with that single commit)
- update PR main message (include GitHub issue number, plus any comments you want to share about your implementation - this will be really valuable some 4-5 weeks from now, should we need to revisit anything)
